src/triedb.rs

```diff
@@ -147,6 +149,56 @@ impl<'db> Trie for TrieDB<'db> {
     }
 }
 
+impl<'db> CryptoStructure for TrieDB<'db> {
+    fn make_proof(&self, key: &H256) -> crate::Result<(CryptoProofUnit, CryptoProof)> {
+        type Unit = CryptoProofUnit;
+
+        fn make_proof_upto<'k>(
+            db: &'k dyn HashDB,
+            path: &'k NibbleSlice<'_>,
+            hash: &'k H256,
+        ) -> crate::Result<(Option<Bytes>, Vec<Bytes>)> {
+            let node_rlp = db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(*hash))?;
+
+            match Node::decoded(&node_rlp) {
+                Some(Node::Leaf(partial, value)) => {
+                    if &partial == path {
+                        Ok((Some(value.to_vec()), vec![node_rlp]))
+                    } else {
+                        Ok((None, vec![node_rlp]))
+                    }
+                }
+                Some(Node::Branch(partial, children)) => {
+                    if path.starts_with(&partial) {
+                        let left_path = path.mid(partial.len() + 1);
+                        match children[path.mid(partial.len()).at(0) as usize] {
+                            Some(x) => {
+                                let r = make_proof_upto(db, &left_path, &x)?;
```

Comment:
This code creates two unnecessary vectors per branch.
```rust
let (value, mut reversed_proof) = make_proof_upto(db, &left_path, &x)?;
reversed_proof.push(node_rlp);
Ok((r.0, reversed_proof))
```
If you make `make_proof_upto` returns a reversed proof, you can make it more efficient.